pkg/istio/onboarding/controller.go

```diff
@@ -0,0 +1,246 @@
+func (c *Controller) addServices(services []string, clusterRbacConfig *v1alpha1.RbacConfig) {
+	for _, service := range services {
+		clusterRbacConfig.Inclusion.Services = append(clusterRbacConfig.Inclusion.Services, service)
+	}
+}
+
+// deleteService will delete a service from the ClusterRbacConfig object
+func (c *Controller) deleteServices(services []string, clusterRbacConfig *v1alpha1.RbacConfig) {
+	for _, service := range services {
+		var indexToRemove = -1
+		for i, svc := range clusterRbacConfig.Inclusion.Services {
+			if svc == service {
+				indexToRemove = i
+				break
+			}
+		}
+
+		if indexToRemove != -1 {
+			clusterRbacConfig.Inclusion.Services = removeIndexElement(clusterRbacConfig.Inclusion.Services, indexToRemove)
+		}
+	}
+}
+
+// createClusterRbacConfig creates the ClusterRbacConfig object in the cluster
+func (c *Controller) createClusterRbacConfig(services []string) error {
+	config := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:    model.ClusterRbacConfig.Type,
+			Name:    model.DefaultRbacConfigName,
+			Group:   model.ClusterRbacConfig.Group + model.IstioAPIGroupDomain,
+			Version: model.ClusterRbacConfig.Version,
+		},
+		Spec: &v1alpha1.RbacConfig{
+			Mode: v1alpha1.RbacConfig_ON_WITH_INCLUSION,
+			Inclusion: &v1alpha1.RbacConfig_Target{
+				Services: services,
+			},
+		},
+	}
+
+	_, err := c.store.Create(config)
+	return err
+}
+
+// getServiceList extracts all services from the indexer with the authz
+// annotation set to true.
+func (c *Controller) getServiceList() []string {
+	cacheServiceList := c.serviceIndexInformer.GetIndexer().List()
+	serviceList := make([]string, 0)
+
+	for _, service := range cacheServiceList {
+		svc, ok := service.(*v1.Service)
+		if !ok {
+			log.Println("Could not cast to service object, skipping service list addition...")
+			continue
+		}
+
+		key, exists := svc.Annotations[authzEnabledAnnotation]
+		if exists && key == authzEnabled {
+			serviceName := svc.Name + "." + svc.Namespace + "." + c.dnsSuffix
+			serviceList = append(serviceList, serviceName)
+		}
+	}
+
+	return serviceList
+}
+
+// sync decides whether to create / update / delete the ClusterRbacConfig
+// object based on the current onboarded services in the cluster
+func (c *Controller) sync() error {
+	serviceList := c.getServiceList()
+	config := c.store.Get(model.ClusterRbacConfig.Type, model.DefaultRbacConfigName, "")
+	if config == nil && len(serviceList) == 0 {
+		log.Println("Service list is empty and cluster rbac config does not exist, skipping sync...")
+		return nil
+	}
+
+	if config == nil {
+		log.Println("Creating cluster rbac config...")
+		return c.createClusterRbacConfig(serviceList)
+	}
+
+	clusterRbacConfig, ok := config.Spec.(*v1alpha1.RbacConfig)
+	if !ok {
+		return errors.New("Could not cast to ClusterRbacConfig")
+	}
+
+	if clusterRbacConfig.Inclusion == nil {
+		return errors.New("ClusterRbacConfig inclusion service list is empty")
```

Comment:
is this an error condition, in the case where the crc resource exists, but not the `Inclusion` field, this would return an error... if there are services to be onboarded, we could add the inclusion field and then process the service list

Reply:
That's a good point, it could still exist and we should add service to the object in those cases. I have removed this if statement for those use cases.